src/core/meshcop/dataset_manager.cpp

```diff
@@ -885,6 +885,24 @@ PendingDatasetManager::PendingDatasetManager(Instance &aInstance)
 {
 }
 
+Error PendingDatasetManager::ReadActiveTimestamp(Timestamp &aTimestamp) const
+{
+    Error   error = kErrorNotFound;
+    Dataset dataset;
+
+    // If the delay timer is not running, then the pending dataset is not valid, avoid
+    // reading from nonvolatile memory if there's no point.
+    VerifyOrExit(mDelayTimer.IsRunning());
```

Comment:
Thanks for all the updates.

I noticed this optimization check, but I'm not sure if the assumption (that the delay timer would be running when we have a Pending Dataset) always holds true. For example, after a device reboot, we could have a saved pending dataset (in non-volatile settings) and while the device is still in the `kRoleDetached` or `kRoleDisabled` states, the delay timer might not be running yet. Since `ReadActiveTimestamp()` looks similar to other `Read<Tlv>` methods, someone might use it in these states without realizing it would apply an additional `VerifyOrExit(mDelayTimer.IsRunning())` check.

My suggestion is to remove this check from this method, such that `ReadActiveTimestamp()` simply reads the active timestamp if any dataset exists, regardless of the delay timer's state.

Instead, we could consider adding such a check in `Mle::HandleAnnounce` where we decide whether to ignore a received announce.  We could check if we have a pending dataset, can read its active timestamp, and the pending dataset delay timer is active (indicating we will soon adopt the pending dataset), then, we proceed to compare the timestamps to decide whether or not to ignore the MLE announce.

Thoughts? 

----

Unrelated to this PR, I'm curious about how the following situation should be handled: Suppose we have a pending dataset, but before the delay timer is reached and it is applied, the device is rebooted.

I know that upon reboot, the device will first try to attach using its saved Active Dataset. If it fails, it will try using any saved Pending Dataset (assuming maybe the network has moved to pending one already). 

But if the device is successful in attaching using its saved Active Dataset, what will it do with any saved Pending Dataset? Checking the code, I see that on the Leader, `PendingDatasetManager::StartLeader()` is called, which will start the delay timer. But I don't think the delay timer is started on non-Leader nodes. This may be something to investigate and update in the code. I'm not sure if the implemented behavior aligns with the spec or if we may need any clarifications in the spec itself. @jwhui any thoughts/comments on this?

Reply:
In response to point 1, I've made some updates that I think accomplishes this in a reasonable way, including the one minute cap on blocking announce messages that was discussed in the last working group meeting.

-
For point 2, I thought experimentally I saw nodes here starting their delay timers in that case but I haven't looked closely. That's something I'd like to understand better as well and I can take a look. Suffice for now that if those nodes didn't start delay timers, the announce messages will "clean up" the situation.